gtsam/geometry/Cal3Fisheye.h

```diff
@@ -0,0 +1,211 @@
+ *   [x_d; y_d] = (th_d / r)*[x_i; y_i]
+ * Pixel coordinates:
+ *   K = [fx s u0; 0 fy v0 ;0 0 1]
+ *   [u; v; 1] = K*[x_d; y_d; 1]
+ */
+class GTSAM_EXPORT Cal3Fisheye {
+ protected:
+  double fx_, fy_, s_, u0_, v0_;  // focal length, skew and principal point
+  double k1_, k2_, k3_, k4_;      // fisheye distortion coefficients
+
+ public:
+  enum { dimension = 9 };
+  typedef boost::shared_ptr<Cal3Fisheye>
+      shared_ptr;  ///< shared pointer to fisheye calibration object
+
+  /// @name Standard Constructors
+  /// @{
+
+  /// Default Constructor with only unit focal length
+  Cal3Fisheye()
+      : fx_(1), fy_(1), s_(0), u0_(0), v0_(0), k1_(0), k2_(0), k3_(0), k4_(0) {}
+
+  Cal3Fisheye(const double fx, const double fy, const double s, const double u0,
+              const double v0, const double k1, const double k2,
+              const double k3, const double k4)
+      : fx_(fx),
+        fy_(fy),
+        s_(s),
+        u0_(u0),
+        v0_(v0),
+        k1_(k1),
+        k2_(k2),
+        k3_(k3),
+        k4_(k4) {}
+
+  virtual ~Cal3Fisheye() {}
+
+  /// @}
+  /// @name Advanced Constructors
+  /// @{
+
+  Cal3Fisheye(const Vector& v);
+
+  /// @}
+  /// @name Standard Interface
+  /// @{
+
+  /// focal length x
+  inline double fx() const { return fx_; }
+
+  /// focal length x
+  inline double fy() const { return fy_; }
+
+  /// skew
+  inline double skew() const { return s_; }
+
+  /// image center in x
+  inline double px() const { return u0_; }
+
+  /// image center in y
+  inline double py() const { return v0_; }
+
+  /// First distortion coefficient
+  inline double k1() const { return k1_; }
+
+  /// Second distortion coefficient
+  inline double k2() const { return k2_; }
+
+  /// First tangential distortion coefficient
+  inline double k3() const { return k3_; }
+
+  /// Second tangential distortion coefficient
+  inline double k4() const { return k4_; }
+
+  /// return calibration matrix
+  Matrix3 K() const;
+
+  /// return distortion parameter vector
+  Vector4 k() const { return Vector4(k1_, k2_, k3_, k4_); }
+
+  /// Return all parameters as a vector
+  Vector9 vector() const;
+
+  /**
+   * @brief convert intrinsic coordinates [x_i; y_i] to (distorted) image
+   * coordinates [u; v]
+   * @param p point in intrinsic coordinates
+   * @param Dcal optional 2*9 Jacobian wrpt intrinsic parameters (fx, fy, ...,
+   * k4)
+   * @param Dp optional 2*2 Jacobian wrpt intrinsic coordinates (xi, yi)
+   * @return point in (distorted) image coordinates
+   */
+  Point2 uncalibrate(const Point2& p, OptionalJacobian<2, 9> Dcal = boost::none,
+                     OptionalJacobian<2, 2> Dp = boost::none) const;
+
+  /// Convert (distorted) image coordinates [u;v] to intrinsic coordinates [x_i,
+  /// y_i]
+  Point2 calibrate(const Point2& p, const double tol = 1e-5) const;
+
+  /// Derivative of uncalibrate wrpt intrinsic coordinates [x_i; y_i]
+  Matrix2 D2d_intrinsic(const Point2& p) const;
```

Comment:
Why do we need this? I think this might be another pattern that had its use in our old code, but it seems is not needed here. It leads to a lot of copy/paste in the cpp code.